File: Sources/HedosKernel/Jobs/Job.swift
import Foundation

public enum JobState: String, Codable, Hashable, Sendable {
    case queued
    case preparing
    case running
    case done
    case failed
    case cancelled

    public var isTerminal: Bool {
        switch self {
        case .done, .failed, .cancelled: true
        case .queued, .preparing, .running: false
        }
    }
}

public struct JobProgress: Codable, Hashable, Sendable {
    public var fraction: Double
    public var step: Int?
    public var totalSteps: Int?

    public init(fraction: Double = 0, step: Int? = nil, totalSteps: Int? = nil) {
        self.fraction = fraction
        self.step = step
        self.totalSteps = totalSteps
    }

    public static let none = JobProgress()
}

public struct Job: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public let modelID: String
    public let capability: Capability
    public let payload: JSONValue
    public var state: JobState
    public var progress: JobProgress
    public var queueReason: String?
    public var preview: Data? = nil
    public var result: [String]
    public var error: String?
    public var submittedAt: Date
    public var startedAt: Date?
    public var finishedAt: Date?

    public init(
        id: String = UUID().uuidString,
        modelID: String,
        capability: Capability,
        payload: JSONValue,
        state: JobState = .queued,
        progress: JobProgress = .none,
        queueReason: String? = nil,
        preview: Data? = nil,
        result: [String] = [],
        error: String? = nil,
        submittedAt: Date = Date(),
        startedAt: Date? = nil,
        finishedAt: Date? = nil
    ) {
        self.id = id
        self.modelID = modelID
        self.capability = capability
        self.payload = payload
        self.state = state
        self.progress = progress
        self.queueReason = queueReason
        self.preview = preview
        self.result = result
        self.error = error
        self.submittedAt = submittedAt
        self.startedAt = startedAt
        self.finishedAt = finishedAt
    }

    enum CodingKeys: String, CodingKey {
        case id, modelID, capability, payload, state, progress, queueReason
        case result, error, submittedAt, startedAt, finishedAt
    }
}

File: kernel/src/jobs/job.rs
//! The `Job` record and its state/progress value types.

use serde::{Deserialize, Serialize};

use crate::records::{Capability, JsonValue};

/// A job's lifecycle state.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum JobState {
    Queued,
    Preparing,
    Running,
    Done,
    Failed,
    Cancelled,
}

impl JobState {
    /// Whether this is an end state (`Done`/`Failed`/`Cancelled`).
    pub fn is_terminal(self) -> bool {
        matches!(
            self,
            JobState::Done | JobState::Failed | JobState::Cancelled
        )
    }
}

/// A job's progress: a `[0, 1]` fraction plus the raw step counts it came from.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct JobProgress {
    pub fraction: f64,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub step: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub total_steps: Option<i64>,
}

impl JobProgress {
    /// Progress at `fraction` derived from `step`/`total_steps`.
    pub fn new(fraction: f64, step: Option<i64>, total_steps: Option<i64>) -> Self {
        Self {
            fraction,
            step,
            total_steps,
        }
    }
}

impl Default for JobProgress {
    fn default() -> Self {
        Self {
            fraction: 0.0,
            step: None,
            total_steps: None,
        }
    }
}

/// A discrete unit of work with progress and a persisted terminal result. Its
/// on-disk form (in `jobs.json`) is internal, so field names are the Rust-native
/// snake_case, not the Swift app's keys.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Job {
    pub id: String,
    pub model_id: String,
    pub capability: Capability,
    pub payload: JsonValue,
    pub state: JobState,
    pub progress: JobProgress,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub queue_reason: Option<String>,
    /// The latest preview frame. Held in memory only — never persisted.
    #[serde(skip)]
    pub preview: Option<Vec<u8>>,
    #[serde(default)]
    pub result: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub error: Option<String>,
    pub submitted_at: i64,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub started_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub finished_at: Option<i64>,
}

impl Job {
    /// A freshly-queued job submitted at `submitted_at` (epoch milliseconds).
    pub fn new(
        id: impl Into<String>,
        model_id: impl Into<String>,
        capability: Capability,
        payload: JsonValue,
        submitted_at: i64,
    ) -> Self {
        Self {
            id: id.into(),
            model_id: model_id.into(),
            capability,
            payload,
            state: JobState::Queued,
            progress: JobProgress::default(),
            queue_reason: None,
            preview: None,
            result: Vec::new(),
            error: None,
            submitted_at,
            started_at: None,
            finished_at: None,
        }
    }
}
